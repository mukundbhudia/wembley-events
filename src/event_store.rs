use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

use chrono::*;
use icalendar::*;

use crate::{SerpapiEvents, WembleyEvent};

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct WembleyEvents {
    pub events: BTreeMap<usize, WembleyEvent>,
}

impl WembleyEvents {
    pub fn new() -> Self {
        WembleyEvents {
            events: BTreeMap::new(),
        }
    }

    pub fn get_events(&self) -> &BTreeMap<usize, WembleyEvent> {
        &self.events
    }

    pub fn build_events_from_html(mut self, html: String) -> WembleyEvents {
        let serp_api_events = serde_json::from_str::<SerpapiEvents>(&html).unwrap_or_default();

        let mut year = serp_api_events.search_metadata.created_at;
        year.truncate(4);

        self.events = serp_api_events
            .events_results
            .into_iter()
            .map(|mut e| {
                let mut swapped_date = e.date.start_date.split_whitespace().collect::<Vec<&str>>();
                swapped_date.push(&year);
                swapped_date.swap(0, 1); // swap [month, day] to [day, month]

                e.date.start_date = swapped_date.join(" ");
                e.into()
            })
            .enumerate()
            .collect::<BTreeMap<usize, WembleyEvent>>();

        Self {
            events: self.events,
        }
    }

    pub fn build_calendar_from_events(self) -> Calendar {
        let mut calendar = Calendar::new();

        self.events.into_iter().for_each(|(_, event)| {
            if let Some(ymd) = event.date_to_ymd() {
                let wembley_event = Event::new()
                    .all_day(Utc.ymd(ymd.year, ymd.month, ymd.day))
                    .summary(&event.title)
                    .description(&event.description)
                    .done();

                calendar.push(wembley_event);
            };
        });

        calendar
    }

    pub fn build_json_from_events(self) -> String {
        serde_json::to_string(&self.events.values().collect::<Vec<_>>()).unwrap()
    }
}

impl Default for WembleyEvents {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use crate::{serpapi_test_output_json_1, serpapi_test_output_json_2};

    use super::*;

    #[test]
    fn build_events_from_html() {
        let body = serpapi_test_output_json_1();
        let wembley_events = WembleyEvents::new().build_events_from_html(body);

        assert_eq!(wembley_events.get_events().len(), 10);
    }

    #[test]
    fn build_calendar_from_events() {
        let body = serpapi_test_output_json_1();
        let wembley_events = WembleyEvents::new().build_events_from_html(body);

        let calendar = wembley_events.build_calendar_from_events();

        assert_eq!(calendar.len(), 10);
    }

    #[test]
    fn check_events_match_calendar() {
        let body = serpapi_test_output_json_1();
        let wembley_events = WembleyEvents::new().build_events_from_html(body);
        let calendar_built_from_events = wembley_events.build_calendar_from_events();

        insta::assert_debug_snapshot!(calendar_built_from_events);
    }

    #[test]
    fn check_events_match_calendar_with_blank_html() {
        let body = serpapi_test_output_json_2();
        let wembley_events = WembleyEvents::new().build_events_from_html(body);
        let calendar_built_from_events = wembley_events.build_calendar_from_events();

        insta::assert_debug_snapshot!(calendar_built_from_events);
    }

    #[test]
    fn check_events_match_calendar_with_blank_string() {
        let body = String::new();
        let wembley_events = WembleyEvents::new().build_events_from_html(body);
        let calendar_built_from_events = wembley_events.build_calendar_from_events();

        insta::assert_debug_snapshot!(calendar_built_from_events);
    }

    #[test]
    fn check_events_match_calendar_json() {
        let body = serpapi_test_output_json_1();
        let wembley_events_as_json = WembleyEvents::new()
            .build_events_from_html(body)
            .build_json_from_events();

        insta::assert_json_snapshot!(wembley_events_as_json);
    }
}
