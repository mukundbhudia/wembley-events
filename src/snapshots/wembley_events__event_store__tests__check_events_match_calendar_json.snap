---
source: src/event_store.rs
assertion_line: 139
expression: wembley_events_as_json

---
"[{\"date\":\"19 Jun 2022\",\"time\":\"Sun, Jun 19\",\"place\":\"Club Wembley\",\"title\":\"Harry Styles\",\"description\":\"Multi-platinum recording artist Harry Styles announces the rescheduled dates for his world tour, in addition to new shows added across the globe. Styles will kick off his colossal 32-city outing...\"},{\"date\":\"12 Jun 2022\",\"time\":\"Sun, 2 PM\",\"place\":\"Wembley Stadium\",\"title\":\"David Guetta @ Wembley Stadium\",\"description\":\"Capital's Summertime Ball\"},{\"date\":\"5 Jun 2022\",\"time\":\"Tomorrow, 1 – 10 PM\",\"place\":\"BOXPARK Wembley\",\"title\":\"Soulful Sundays\",\"description\":\"Join BOXPARK Wembley and their resident DJ's for all things soulful\"},{\"date\":\"29 Jun 2022\",\"time\":\"Wed, 4:30 – 8:30 PM\",\"place\":\"Wembley Stadium\",\"title\":\"Ed Sheeran @ Wembley Stadium\",\"description\":\"span Ed Sheeran has announced the first leg of his ‘+ - = ÷ x Tour’ (pronounced ‘The Mathematics Tour’) , taking place in stadiums throughout 2022. Kicking off in April next year, the tour will...\"},{\"date\":\"25 Jun 2022\",\"time\":\"Sat, 5 – 9 PM\",\"place\":\"Wembley Stadium\",\"title\":\"Dylan @ Wembley Stadium\",\"description\":\"span Ed Sheeran has announced the first leg of his ‘+ - = ÷ x Tour’ (pronounced ‘The Mathematics Tour’) , taking place in stadiums throughout 2022. Kicking off in April next year, the tour will...\"},{\"date\":\"16 Aug 2022\",\"time\":\"Tue, 5 – 9 PM\",\"place\":\"Wembley Stadium\",\"title\":\"Coldplay @ Wembley Stadium\",\"description\":\"Global super pop rock band Coldplay will be hitting the road on tour in 2022 and will be performing live on stage in their Music of the Spheres World Tour ! This particular live concert is...\"},{\"date\":\"24 Jun 2022\",\"time\":\"Fri, 5 – 9 PM\",\"place\":\"Wembley Stadium\",\"title\":\"Ed Sheeran + - = ÷ x Tour\",\"description\":\"Ed Sheeran Seated Ticket & Hotel Experience This Experience Includes: • Reserved Seated Ticket (Block 104 or 141) Tickets seated together • One night hotel stay including complimentary breakfast •...\"},{\"date\":\"20 Aug 2022\",\"time\":\"Sat, 5 – 9 PM\",\"place\":\"Wembley Stadium\",\"title\":\"Coldplay: Music of the Spheres World Tour\",\"description\":\"Global super pop rock band Coldplay will be hitting the road on tour in 2022 and will be performing live on stage in their Music of the Spheres World Tour ! This particular live concert is...\"},{\"date\":\"26 Jun 2022\",\"time\":\"Sun, 7 – 9 PM\",\"place\":\"OVO Arena\",\"title\":\"Jubin Nautiyal - RESCHEDULED\",\"description\":\"Jubin Nautiyal at OVO Arena Wembley in Wembley Park\"},{\"date\":\"10 Jun 2022\",\"time\":\"Fri, 7:30 – 11:00 PM\",\"place\":\"The SSE Arena, Wembley\",\"title\":\"50 Cent\",\"description\":\"SJM Concerts 50 CENT FRI, 10 JUN 2022 at 07:30PM BST Doors Open: 06:00PM OnSale: Fri, 28 Jan 2022 at 09:30AM GMT Announcement: Wed, 26 Jan 2022 at 10:00AM GMT Hip Hop heavyweight 50 Cent, has...\"}]"
